mex/backend/security.py

```diff
@@ -1,72 +1,98 @@
 from typing import Annotated
 
 from fastapi import Depends, HTTPException
-from fastapi.security import APIKeyHeader
+from fastapi.security import APIKeyHeader, HTTPBasic, HTTPBasicCredentials
 from starlette import status
 
 from mex.backend.settings import BackendSettings
-from mex.backend.types import APIKey
+from mex.backend.types import APIKey, APIUserPassword
 
 X_API_KEY = APIKeyHeader(name="X-API-Key", auto_error=False)
+X_API_CREDENTIALS = HTTPBasic(auto_error=False)
 
 
-def has_write_access(api_key: Annotated[str | None, Depends(X_API_KEY)]) -> None:
-    """Verify if api key has write access.
+def has_write_access(
+    api_key: Annotated[str | None, Depends(X_API_KEY)] = None,
+    credentials: Annotated[
+        HTTPBasicCredentials | None, Depends(X_API_CREDENTIALS)
+    ] = None,
+) -> None:
+    """Verify if provided api key or credentials have write access.
 
     Raises:
-        HTTPException if no header is provided or APIKey does not have write access.
+        HTTPException if no header or provided APIKey/credentials have no write access.
 
     Args:
         api_key: the API key
+        credentials: username and password
 
     Settings:
-        backend_user_database: checked for presence of api_key
+        check credentials in backend_user_database or backend_api_key_database
     """
-    if not api_key:
+    if not api_key and not credentials:
         raise HTTPException(
             status_code=status.HTTP_401_UNAUTHORIZED,
-            detail="Missing authentication header X-API-Key.",
+            detail="Missing authentication header X-API-Key or credentials.",
         )
 
     settings = BackendSettings.get()
-    user_database = settings.backend_user_database
-    can_write = APIKey(api_key) in user_database.write
+    can_write = False
+    if api_key:
+        api_key_database = settings.backend_api_key_database
+        can_write = APIKey(api_key) in api_key_database.write
+    elif credentials:
+        api_write_user_db = settings.backend_user_database.write
+        user, pw = credentials.username, credentials.password
+        can_write = APIUserPassword(pw) == api_write_user_db.get(user)
     if not can_write:
         raise HTTPException(
             status_code=status.HTTP_403_FORBIDDEN,
-            detail="Unauthorized API Key.",
+            detail=f"Unauthorized {'API Key' if api_key else 'credentials'}.",
         )
```

Comment:
for basic auth to play nice with browsers ( = prompt you to input your credentials),
we need to return [`headers={"WWW-Authenticate": "Basic"}`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate) on all 401 and 403s.

however, it would be nice to only set that header if the request came from a browser and not for api-clients (because they should use the x-api-key header, not basic auth).
to differentiate, we could use a new dependency
`user_agent: Annotated[str, Header(include_in_schema=False)] = "n/a"`
and check if the user_agent starts with `"Mozilla"` (yes, all browser user-agents start with Mozilla. [it is a long and sad story.](https://stackoverflow.com/a/1114297/2216759))

Reply:
Thanks for pointing me to how to solve this, I tried it here: https://github.com/robert-koch-institut/mex-backend/pull/18/commits/ebad49b0847add353c8bcf18934ea126b7056596